### ISSUE TITLE
Poor INSERT performance with ExecuteNonQuery() and many parameters

### ISSUE THREAD

**Issue by KristianWedberg** [dotnet/SqlClient#974]
### Description

When inserting rows with `ExecuteNonQuery()` and placeholder parameters, performance 
(measured as **inserted parameters per second**) with the `Microsoft.Data.SqlClient` v2.1.1  
and `System.Data.SqlClient` providers is progressively impacted from about 50+ parameters 
per statement, both for single rows and for multirow batches.

Inserting with 2048 parameters in the statement is up to 7 times slower than with 
128 parameters.

The issue is _not_ present when using the `System.Data.Odbc` provider (or non-SQL Server
providers & databases), and performance is up to 25 times slower (when using 2048 parameters 
per statement) than with the ODBC provider.

I've tested with Windows 10, .NET5.0 and a local SQL Server 2019 database, varying the 
number of columns from 1 to 1024, and number of rows in each batch from 1 to 1000.

Note: The 'waviness' at 1000 and 2000 parameters is just an effect of the 1000/1024 and 2000/2048 
numbers being separated in the graph - see the write-up for details:

![Average Throughput](https://github.com/KristianWedberg/sql-batch-insert-performance/raw/main/average-throughput.png)

The target table is un-indexed and the inserts uses SQL-92 syntax:

```
INSERT INTO tablename (column-a, [column-b, ...])
VALUES (@r0_c0, [@r0_c1, ...]),
       (@r1_c0, [@r1_c1, ...]),
       ...
```

While table valued parameters and bulk inserts don't have this issue, they won't 
help when inserting a single or just a few very wide rows at a time, so it would be 
really useful to have this addressed (or identify any gremlins on my part!)

Please see the full write-up with charts, tables, BenchmarkDotNet info etc. at:

https://github.com/KristianWedberg/sql-batch-insert-performance

### Reproduce

Fully runnable source (using BenchmarkDotNet) and instructions at:

https://github.com/KristianWedberg/sql-batch-insert-performance

### Expected behavior

Using more parameters (beyond 128) in an insert row or batch should increase throughput, 
measured as the number of parameters inserted per second, just like it does with 
`System.Data.Odbc` and other non-SqlClient providers.

### Technical details

- Microsoft.Data.SqlClient Version="2.1.1"
- System.Data.Odbc Version="5.0.0"
- System.Data.SqlClient Version="4.8.2"
- OS=Windows 10.0.19042
- .NET Core SDK=5.0.103

**Comment by Wraith2** [comment]
What is the real world use case for using so many parameters?

**Comment by KristianWedberg** [comment]
I've personally come across this with:

- Wide rows to capture different stages of a process (think lots of dates and associated data). One real estate company used 450 columns for this.
  - 'Type 3 slowly changing dimensions' is a similar use case
- Multi-row batches to improve performance

**Comment by Wraith2** [comment]
> Multi-row batches to improve performance
While table valued parameters and bulk inserts can handle this use case for SQL Server, those techniques are not portable, which is a big draw-back

This is likely to be addressed in .net 6 for most major providers with https://github.com/dotnet/runtime/issues/28633 

I'll try to take a look at this when I can get some time but I make no promises.

**Comment by KristianWedberg** [comment]
@roji I'm very much looking forward to the upcoming .NET6 batch API. That said:

- It's still many years before we can let go of all <.NET6 frameworks
- The batch API won't help with wide single row inserts

**Comment by roji** [comment]
> It's still many years before we can let go of all <.NET6 frameworks

This is very true, though it's probably going to be possible (and advisable) to expose the batching API even for older TFMs - just without implementing the ADO.NET base classes. This would make the API accessible regardless of which TFM you're using (possibly even .NET Framework); that's my plan for Npgsql/PostgreSQL, in any case.

> The batch API won't help with wide single row inserts

This is true - and at that point it's also a question of how common it is to have such a large number of columns etc.

**Comment by Wraith2** [comment]
Just had a quick look. You're using System.Data.SqlClient. Can you try it with Microsoft.Data.SqlClient from this repository and see if the numbers are any better to start with?

**Comment by KristianWedberg** [comment]
I'm already using both `Microsoft.Data.SqlClient` (the "Microsoft" or "MicrosoftSync" results) and `System.Data.SqlClient` in the code and in the results. Their performance are identical.

**Comment by Wraith2** [comment]
Ah, I see it now. Had to do some fiddling to profile it and broke something. My mistake.

I've found some interesting areas that I was already aware of but were a way down my list of things to investigate. It's mostly string composition and array resizing overhead. As the parameter count goes up it spends more time allocating and re-allocating to grow and there's now a lot the caller can do about that. It can be reworked to avoid intermediate strings and look at pre-sizing of string builders and target arrays in a few places.

Two things stand out that you could do for the benchmark. Can you keep pre-boxed ints that are used as parameter values around? Their boxing and unboxing is a big source of GC activity. I know that isn't reflectve of real world data but in real world scenarios if you have strongly typed input data of limited or frequently known ranges (bool, small ints) keeping cached boxed versions of some values around and using those will gain you some perf.

If possible reusing SqlParameter instances would be good. They're quite large and you're using a lot of them.

**Comment by Wraith2** [comment]
That just means the individual benches aren't running for long enough to be affected by memory issues which is a good thing in general but if you want to run these tasks for a long time preboxing and reusing things will save you some memory. There being no throughput difference isn't surprising I was just making observations. The current limiting factors are parallelism and buffer resizing as I said.

**Comment by Wraith2** [comment]
I made the code execute about 20% faster and it made no difference to throughput. The code isn't the bottleneck, the network is. ODBC just appears to be more efficient about how the parameters are sent somehow. Since i can't change what needs to be put on the wire or how long it takes to send and receive the messages over the network there is no clear way to improve performance here.

**Comment by roji** [comment]
@Wraith2 interesting if this is actually a wire protocol data change - it may be worth comparing what they're doing with wireshark. If SqlClient is using TDS in an inefficient way that might be another thing to track somewhere...

**Comment by Wraith2** [comment]
Sure. One of many many things to investigate at some point. 
I'll keep the branch with my changes in around and try to PR them at some point because someone may appreciate the additional speed in some scenario, it just doesn't help in this one. 

My advice on this one would be to get closer to the sql server and attempt parallel imports.

**Comment by cheenamalhotra** [comment]
You can also try to compare how queries execute on server side by comparing SQL Server Profiler traces. That will tell you whether parameter metadata caching is adding benefit or not on client side.

**Comment by KristianWedberg** [comment]
@Wraith2 When I run this single test case (and with `_scaleNumberOfStatements = 40_000`, `.AddJob(Job.Dry)`, 
again with a _**local**_ database), I get the following CPU utilization:

| provider | RPS | CPR | NOS | Mean | Throughput [params/s] | Client CPU | SQL Server CPU |
|--------- |----------------- |-------------- |------------------- |--------- |--------:|------:|---:|
| Microsoft | 32 | 64 |   616 | 26.39 s |   47812 | 1% | 30% |
|      Odbc | 32 | 64 | 17921 | 33.12 s | 1108118 | 15% | 30% |

Certainly not proven, but it looks like:

- In both cases the SQL Server process is likely bottlenecked on a single thread (I have 4 cores total).
- The Microsoft provider is (as you said) not a CPU bottleneck, since the client side only consumes 1%.
- The ODBC test case is performing 29 times more work than the Microsoft test case, so it makes perfect 
  sense that it has ~15 times higher client side CPU utilization.
  
The question is - why does the SQL Server process consume so much CPU when it receives these statements
with 2048 parameters (and starts to be performance limited at 50+ parameters)? 

Can this be raised with the SQL Server team?

**Comment by Wraith2** [comment]
I can provide you with the dev nuget package for my build if you want to try it out and verify the perf difference for yourself.

**Comment by KristianWedberg** [comment]
Some more tests:

- System.Data.SqlClient 4.4.0 (from 2017) and the latest Microsoft.Data.SqlClient 2.1.2 both showed the same issue.
- SQL Server 2012 and SQL Server 2016 both showed the same issue.

So it's not a recent issue.

**Comment by Wraith2** [comment]
For shits and giggles you can give this a try and see that it changes nothing if you like. 
[Microsoft.Data.SqlClient.3.0.0-dev.zip](https://github.com/dotnet/SqlClient/files/6135393/Microsoft.Data.SqlClient.3.0.0-dev.zip)

**Comment by KristianWedberg** [comment]
@Wraith2 Thanks. As expected by now, Microsoft.Data.SqlClient.3.0.0-dev.zip made no difference whatsoever.

**Comment by KristianWedberg** [comment]
Profiling shows that query plans (including operator values) look identical: T-SQL Insert <-- Table Insert <-- Constant Scan

The SqlClient and Odbc queries look identical, except for SqlClient using named instead of positional parameters (here with only 3 columns and 2 rows). The parameters being named might well be the reason for the performance difference.

SqlClient:
```
exec sp_executesql N'INSERT INTO insertbenchmark_2_3_MicrosoftSync (d0,d1,d2) 
    VALUES (@R0_C0,@R0_C1,@R0_C2),(@R1_C0,@R1_C1,@R1_C2)',N'@R0_C0 int,@R0_C1 int,@R0_C2 int,@R1_C0 int,@R1_C1 int,@R1_C2 int',
    @R0_C0=0,@R0_C1=1,@R0_C2=2,@R1_C0=3,@R1_C1=4,@R1_C2=5
```

Odbc:
```
exec sp_executesql N'INSERT INTO insertbenchmark_2_3_OdbcSync      (d0,d1,d2) 
    VALUES (@P1,@P2,@P3),(@P4,@P5,@P6)',N'@P1 int,@P2 int,@P3 int,@P4 int,@P5 int,@P6 int',
    0,1,2,3,4,5
```

Interestingly, when running with many (in this case 32 * 64 = 2048) parameters, SqlClient reports both CPU consumption, physical writes and duration for _every_ query, while Odbc only shows CPU consumption and duration for the _first_ query and _no_ physical writes for _any_ query.

SqlClient:

| CPU (ms) | Logical Reads | Physical Writes | Duration (ms) |
| ---: | ---: | ---: | ---: |
| 94 | 96 | 2 | 95 |
| 47 | 42 | 1 | 40 |
| 47 | 42 | 1 | 42 |
| 47 | 42 | 1 | 43 |
| 46 | 43 | 1 | 43 |
| 47 | 42 | 1 | 45 |
| 32 | 42 | 1 | 42 |


Odbc:

| CPU | Reads | Writes | Duration |
| ---: | ---: | ---: | ---: |
| 63 | 96 | 0 | 55 |
| 0 | 42 | 0 | 0 |
| 0 | 43 | 0 | 0 |
| 0 | 42 | 0 | 0 |
| 0 | 42 | 0 | 0 |
| 0 | 43 | 0 | 0 |
| 0 | 42 | 0 | 0 |

So possibly the named parameters make SQL Server consume significant CPU on every query, whereas with positional parameters 
that work is either not needed or a cached copy is used.

I don't know how to troubleshoot this further into SQL Server, do any of you have a handle on that?

**Comment by Wraith2** [comment]
I think you'll have to open a support case for that.

Once we get to the point where batching is possible would you be interested in trying out the preview builds for it? This is a use case that should be able to benefit from it and it would be useful to have some idea how it will react.

**Comment by Wraith2** [comment]
```sql
exec sp_executesql N'INSERT INTO insertbenchmark_2_3_MicrosoftSync (d0,d1,d2) 
    VALUES (@R0_C0,@R0_C1,@R0_C2),(@R1_C0,@R1_C1,@R1_C2)',N'@R0_C0 int,@R0_C1 int,@R0_C2 int,@R1_C0 int,@R1_C1 int,@R1_C2 int',
    @R0_C0=0,@R0_C1=1,@R0_C2=2,@R1_C0=3,@R1_C1=4,@R1_C2=5
```
If you just changed the last line of @Name=value pairs to a value list without changing the values clauses would it still work? So the question is if the names are important if the names aren't specified in the value list. If it's still valid syntax (indicating @P\d is not special) then I could quirk the BuildParamList to just not output the names if an appcontext flag were set.

**Comment by ErikEJ** [comment]
@Wraith2 Based on my findings, I think that would be an interesting experiment.

**Comment by Wraith2** [comment]
Ok, give it a go then. Make sure you clear out your nuget package cache of previous versions of 3.0.0-dev before you try. Set the switch with `AppContext.SetSwitch("Switch.Microsoft.Data.SqlClient.UseODBCParameterFormat",true);` before you you run any queries because it'll cache the value in a static.
[Microsoft.Data.SqlClient.3.0.0-dev.zip](https://github.com/dotnet/SqlClient/files/6150080/Microsoft.Data.SqlClient.3.0.0-dev.zip)

It's not a string concat it's actual protocol level but there's an option to just send 0 as the name length and it seems to work under the debugger, I get the correct rowsaffected back.

**Comment by KristianWedberg** [comment]
@Wraith2  I think me and @ErikEJ have too similar avatars :-)

Correct me if I'm wrong, but I don't think that'll will work particularly well in practice: 

The switch will make the whole provider into a positional parameter provider, so _all_ uses of the provider will have to use positional parameters, not just this insert. That means e.g. that other queries that use the same parameter name in multiple places in a query would need to expand and reorder parameter values to match the order and number of named parameters in the query. I'm assuming your switch won't do that?

**Comment by Wraith2** [comment]
You and Erik do look sort of similar.. though you're more tanned :grin:. But no I didn't mix you up since erik responded that it might be a good idea I thought I'd give it a try. If we find something that helps at this level then you aren't waiting for a new version of SQL Server for improvements.

If we can prove that using the alternative format avoids the speed problem then a case could be made for adding public surface area to the SqlCommand class to change the style on a per-command basis, if this test has no effect then there would be no point doing it.

Yes it is a global switch. The names are still sent as part of the parameter list but they aren't sent as part of the value list. So name matching might still work, does using the odbc driver allow the use of named parameters? If you get chance could you run the numbers through your benches and see if this is a dead end?

**Comment by KristianWedberg** [comment]
Gotcha, will try it out.

ODBC is positional parameters only.

Also, I'm certainly happy to try out the batching API previews.

**Comment by ErikEJ** [comment]
@Wraith2  Just tried the build, still see named parameters in the Profiler trace !?

**Comment by Wraith2** [comment]
That's down to whatever is showing you the trace. The RPC is build and sent with the parameter list as a string but the parameter values are written separately. Effectively I force it into the zero length case of this:
https://github.com/dotnet/SqlClient/blob/c37092a378a972c3799bdf79db6fa58f779ba9fd/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs#L9695-L9709

I could also fiddle about with [SqlCommand.BuildParamList](https://github.com/dotnet/SqlClient/blob/c37092a378a972c3799bdf79db6fa58f779ba9fd/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.cs#L5828) which writes the name list but odbc sends a similar list and it shouldn't matter if the p\d format isn't special. 

What I think is happening is that when you have named params and named values it uses a name lookup which requires comparisons so it slows as the count of comparisons increases, if you have unnamed values you can only do an index lookup which scales linearly as we see with the odbc driver.

**Comment by ErikEJ** [comment]
>> That's down to whatever is showing you the trace.  

Not sure I understand, the ODBC statement never had named values, but the SqlClient has. (as still has)

**Comment by Wraith2** [comment]
The parameter values part, e.g. "@R0_C0=0,@R0_C1=1,@R0_C2=2,@R1_C0=3,@R1_C1=4,@R1_C2=5" isn't sent as a string it's written directly as part of the RPC request. If you're seeing that as a string then something is interpretting it. The parameter names aren't being written for the values.

They are still being written for the parameter list part where the types are present.

**Comment by ErikEJ** [comment]
I am seeing it in the SQL Server profiler. 
From ODBC is see just values. Just like in the sample statement above.

**Comment by Wraith2** [comment]
Sql server profile is interpreting the values into a string to display to you, I didn't put them there, have a look in wireshark perhaps if you want to ensure that. I could remove the parameter names from the parameter list as well and just replace them with positional values but i'm not sure if that's what the difference is, if it is then it implies that it treats P specially.

**Comment by Wraith2** [comment]
In Sql Server Profiler i see the same format albeit it names instead of placeholders. newlines for clarity of argument parts

```sql
exec sp_executesql 
N'INSERT INTO insertbenchmark_4_4_MicrosoftAsync (d0,d1,d2,d3) VALUES (@R0_C0,@R0_C1,@R0_C2,@R0_C3),(@R1_C0,@R1_C1,@R1_C2,@R1_C3),(@R2_C0,@R2_C1,@R2_C2,@R2_C3),(@R3_C0,@R3_C1,@R3_C2,@R3_C3)',
N'@R0_C0 int,@R0_C1 int,@R0_C2 int,@R0_C3 int,@R1_C0 int,@R1_C1 int,@R1_C2 int,@R1_C3 int,@R2_C0 int,@R2_C1 int,@R2_C2 int,@R2_C3 int,@R3_C0 int,@R3_C1 int,@R3_C2 int,@R3_C3 int',
3856,3857,3858,3859,3860,3861,3862,3863,3864,3865,3866,3867,3868,3869,3870,3871
```

```sql
exec sp_executesql 
N'INSERT INTO insertbenchmark_4_4_OdbcSync (d0,d1,d2,d3) VALUES (@P1,@P2,@P3,@P4),(@P5,@P6,@P7,@P8),(@P9,@P10,@P11,@P12),(@P13,@P14,@P15,@P16)',
N'@P1 int,@P2 int,@P3 int,@P4 int,@P5 int,@P6 int,@P7 int,@P8 int,@P9 int,@P10 int,@P11 int,@P12 int,@P13 int,@P14 int,@P15 int,@P16 int',
9376,9377,9378,9379,9380,9381,9382,9383,9384,9385,9386,9387,9388,9389,9390,9391

**Comment by KristianWedberg** [comment]
@Wraith2 I've cleared all package caches, uninstalled the old package and installed the new version, and set the switch at program start. I still get the slow performance and profiler shows the @parameterName=value:

```
exec sp_executesql N'INSERT INTO insertbenchmark_2_3_MicrosoftSync (d0,d1,d2) 
    VALUES (@R0_C0,@R0_C1,@R0_C2),(@R1_C0,@R1_C1,@R1_C2)'
    ,N'@R0_C0 int,@R0_C1 int,@R0_C2 int,@R1_C0 int,@R1_C1 int,@R1_C2 int'
    ,@R0_C0=36,@R0_C1=37,@R0_C2=38,@R1_C0=39,@R1_C1=40,@R1_C2=41
```

Am I missing something to enable the new behaviour?

**Comment by Wraith2** [comment]
There's only the switch that i mentioned. `AppContext.SetSwitch("Switch.Microsoft.Data.SqlClient.UseODBCParameterFormat",true);`

**Comment by KristianWedberg** [comment]
Interesting...

Your modified `M.D.SqlClient` version is _a lot_ faster with 50+ parameters per batch (yay!) and keeps pace with the ODBC provider up to about 256 parameters per batch. With 256+ parameters it hits some other bottleneck and eventually slows down drastically. At 2000 parameters it's 10 times slower than ODBC but still 2-3 times faster than its normal (unmodified) performance.

### Average Throughput [parameters/s]

| Parameters | Microsoft | Odbc | System |
| ---: | ---: | ---: | ---: |
| 1 |	4,169 |	4,255 |	4,367 |
| 2 |	8,739 |	8,709 |	8,787 |
| 4 |	17,036 |	17,032 |	17,338 |
| 8 |	31,844 |	32,845 |	28,418 |
| 16 |	61,067 |	62,204 |	59,588 |
| 32 |	110,930 |	110,828 |	102,149 |
| 64 |	185,495 |	187,999 |	156,634 |
| 128 |	297,391 |	288,301 |	190,994 |
| 256 |	374,091 |	399,894 |	169,225 |
| 512 |	388,626 |	526,077 |	122,313 |
| 1,000 |	294,432 |	609,618 |	69,938 |
| 2,000 |	58,372 |	604,888 |	23,359 |

What's next? The only _great_ solution is still to have the bottlenecks on the SQL Server side removed. We're an ISV and not an end user of SQL Server so don't have a support contract...

**Comment by cheenamalhotra** [comment]
Apart from the length of query getting reduced is there any other significant difference in your changes @Wraith2 ?
I can trigger some talks with SQL team internally as this is very strange but interesting finding.

**Comment by Wraith2** [comment]
The change from yesterday, which is the only one to make a difference, is really simple. Just check for the switch and bypass the parameter name write in TdsParser.WriteParameterName:

```csharp
        private static bool s_omitParameterNames;
        private static bool s_omitParameterNamesChecked;

        private void WriteParameterName(string parameterName, TdsParserStateObject stateObj)
        {
            if (!s_omitParameterNamesChecked)
            {
                AppContext.TryGetSwitch("Switch.Microsoft.Data.SqlClient.UseODBCParameterFormat", out s_omitParameterNames);
                s_omitParameterNamesChecked = true;
            }

            //paramLen
            //paramName
            if (!string.IsNullOrEmpty(parameterName) && !s_omitParameterNames)
            {
                Debug.Assert(parameterName.Length <= 0xff, "parameter name can only be 255 bytes, shouldn't get to TdsParser!");
                int tempLen = parameterName.Length & 0xff;
                stateObj.WriteByte((byte)tempLen);
                WriteString(parameterName, tempLen, 0, stateObj);
            }
            else
            {
                stateObj.WriteByte(0);
            }
        }
```

the parameter names are still sent in the parameter list which might be affecting the parse speed. It's only an extra 4-10 chars per name but that could easily add up. If the speed up we've seen so far is the value matching but the speed still decreases because of the parse it might be worth stubbing in replacement p values instead of names to see how that works out. I can write that later and provide a package.

**Comment by ErikEJ** [comment]
@Wraith2 @cheenamalhotra I noticed that the payload of the sp_executesql command exceeded 64K with 2048 parameters - wonder if that is significant - if so, maybe @KristianWedberg could make the names shorter to investigate?

**Comment by KristianWedberg** [comment]
In the table below:

- "Short" uses the shortest possible parameter name: @p0, @p1, ...
- "Long" uses the longer parameter name: @r0_c0, @r0_c1, ...
- "False" uses `M.D.SqlClient` 3.0.0-dev with `UseODBCParameterFormat` set to `False` (i.e. named parameters)
- "True" uses `M.D.SqlClient` 3.0.0-dev with `UseODBCParameterFormat` set to `True` (i.e. positional parameters)

The data shows that using the shortest possible parameter name improves performance (only) a few percent, both with and without the positional parameter switch enabled.

Put differently, the original large performance reduction (at 50+ or 256+ parameters) compared to ODBC remains.

| Parameters	 | 	Long – False	 | 	Short – False	 | 	Long – True	 | 	Short – True	 |
| ---:	 | 	---:	 | 	---:	 | 	---:	 | 	---:	 |
| 1	 | 	4,094	 | 	4,158	 | 	4,070	 | 	3,963	 |
| 2	 | 	8,496	 | 	8,608	 | 	8,462	 | 	8,372	 |
| 4	 | 	16,287	 | 	16,239	 | 	16,603	 | 	16,433	 |
| 8	 | 	29,966	 | 	31,550	 | 	31,680	 | 	31,802	 |
| 16	 | 	58,808	 | 	58,557	 | 	59,186	 | 	60,130	 |
| 32	 | 	100,190	 | 	101,405	 | 	103,898	 | 	108,711	 |
| 64	 | 	149,093	 | 	157,702	 | 	173,714	 | 	182,843	 |
| 128	 | 	189,947	 | 	203,155	 | 	290,420	 | 	288,445	 |
| 256	 | 	167,050	 | 	180,157	 | 	357,797	 | 	365,522	 |
| 512	 | 	121,819	 | 	128,017	 | 	377,136	 | 	381,154	 |
| 1,000	 | 	69,882	 | 	73,588	 | 	291,245	 | 	292,676	 |
| 2,000	 | 	22,855	 | 	23,086	 | 	58,132	 | 	57,302	 |

**Comment by Wraith2** [comment]
Interesting. With the same param names and format they should be identical as seen in query profiler which indicates that the protocol used to send the query matters. SqlClient is talking TDS and the ODBC driver is using something else but wireshark can pull apart the odbc query directly. That's definitely something that will need the Sql Server team to look at if you can ask them to @cheenamalhotra .

So given that we now know that sending positional parameters can have a performance benefit is this something that is worth putting into this library? /cc @David-Engel 

If it is going to be added it needs to be at SqlCommand level. I see 3 options:
1) Add a new member to SqlCommandBehaviour (`SqlCommandBehaviour.UsePositionalParameters` ?) but doing that would be a breaking change and would be unlikely to make it through api review without a lot of evidence that it would be useful to many people and worth the risk to legacy codebases. I suspect we would need to prove that it has use beyond a single driver and I don't know if postgres or mysql (or oracle but i've never spoken to their maintainers) has this capability or need. ODBC certainly doesn't.
2) Add a new property direction on SqlCommand, not a fan of this but it'll work and is simplest to implement.
3) A slightly shady alternative is to use a free bit in SqlCommandBehaviour and just tell users to or an int constant into the enum and then mask it out internally. This avoids new surface area but is undiscoverable to users and looks dodgy

**Comment by ErikEJ** [comment]
@Wraith2 or just default to positional??? 🥳

**Comment by Wraith2** [comment]
> @Wraith2 or just default to positional??? 🥳

Sure, as long as we can put your email address on that PR so people know how to feedback directly 😁

**Comment by cheenamalhotra** [comment]
I'd say let's also compare execution plans on SQL Server directly through SSMS as that tells us a lot.

All you'd need to do is:

- Paste the sp_execute queries one by one in SSMS and run "Show Execution plan"

That will tell us what server does behind the scenes.

**Comment by ErikEJ** [comment]
I cannot see any difference in the execution plans!
[Plans.zip](https://github.com/dotnet/SqlClient/files/6158030/Plans.zip)

**Comment by cheenamalhotra** [comment]
Ok, Thanks @ErikEJ

I'll trigger talks internally with SQL team today to discuss perf loss in our case and in the mean time we can look at our changes to bring in performance.

We should also compare parametrized queries with MSSQLJDBC as both JDBC and SqlClient drivers are closer knit than ODBC is with SqlClient.

**Comment by ErikEJ** [comment]
Great, MDS should be right up there with the others when it comes to perf!

**Comment by ErikEJ** [comment]
Statistics time on:

Named parameters:

SQL Server parse and compile time: 
   CPU time = 203 ms, elapsed time = 210 ms.

vs

SQL Server parse and compile time: 
   CPU time = 0 ms, elapsed time = 4 ms.

**Comment by Wraith2** [comment]
Measured through management studio or a program?

**Comment by ErikEJ** [comment]
@Wraith2 SSMS

**Comment by v-chojas** [comment]
The ODBC driver also uses TDS, but there is probably a difference in exactly how it submits the query which should be reasonably easy to see in a packet trace.

**Comment by Wraith2** [comment]
I couldn't see anything particularly different. I'm not sure if I turned off encryption on the sqlclient trace. The only thing i notes was that the default odbc packet size was 4K and sql client was 8K.

**Comment by v-chojas** [comment]
Turning off encryption will be necessary to get a readable trace, and also rules out any differences caused by TLS/SSL.

**Comment by Wraith2** [comment]
Ok, so I have traces and there are some differences. Now what do I do with that information as a client? How do I decide which are important and which aren't? For example I'm pretty sure I don't want to change the value of the reset connection header flag because it's important (and yet odbc doesn't use it? should it?) but how about the no metadata option? 

ODBC:
![odbc capture](https://user-images.githubusercontent.com/13322696/111553435-4f3ca100-877c-11eb-9499-ae4ea85c88d8.PNG)

SqlClient:
![sqlclient capture](https://user-images.githubusercontent.com/13322696/111553447-5663af00-877c-11eb-9ce4-3ec6ca64550b.PNG)

I can provide the capture files privately if anyone wants to look at them. I'm not sure where the tiny length discrepancy is coming from because i changed the table names to have the same length.

**Comment by David-Engel** [comment]
> For example I'm pretty sure I don't want to change the value of the reset connection header flag because it's important (and yet odbc doesn't use it? should it?)

ODBC would only have it if the driver manager handed out a previously pooled connection and this was the first query to be executed.

**Comment by Wraith2** [comment]
Both samples were run the same way from the example app provided above. Table setup was done and then I started capturing at a breakpoint and stopped immediately after the query succeeded. the command objects were newed up each time. Looking at a longer running trace the reset behaviour seems to follow what you said. So scratch that one. Next, no metadata?

**Comment by cheenamalhotra** [comment]
As per MS-TDS Specs, in [RPC request](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-tds/619c43b6-9495-4a58-9e49-a4950db245b3), NoMetaData flag is described as:

**fNoMetaData**
_The server sends NoMetaData only if fNoMetadata is set to 1 in the request (see COLMETADATA, section 2.2.7.4)_

Whereas ([COLMETADATA](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-tds/58880b9f-381c-43b2-bf8b-0727a98c4f4c)) where **NoMetaData** is to be received in response does not apply to [RPC Server response](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-tds/a4e706cf-8ae7-4809-87a1-ce293a6f7f68), so it's basically a no-op for SqlClient. You can verify response packets too.

**Comment by Wraith2** [comment]
Manually decoded packets with annotations, download and diff to review.

The differences are:
1) ODBC uses PLP string format not inline, this accounts for the length differences
2) SqlClient sent reset connection flag in the header, already discussed and probably not important
3) SqlClient sent the NoMetadata flag, this is unconditionally set if the command object is in Text mode.

[sqlcient hex.txt](https://github.com/dotnet/SqlClient/files/6163969/sqlcient.hex.txt)
[odbc hex.txt](https://github.com/dotnet/SqlClient/files/6163970/odbc.hex.txt)

**Comment by v-chojas** [comment]
I wouldn't expect to see any difference with 16 parameters, but try looking at one with 512 parameters. That will cause both the SQL and parameter declaration list to be long enough not to fit in a regular nvarchar (4000 chars/8000 bytes). The expression to calculate the length of the parameter list in chars is `8*n + (n-10)*(n>10) + (n-100)*(n>100) + (n-1000)*(n>1000) - 1`, which exceeds 4000 when n exceeds 411.

**Comment by Wraith2** [comment]
What difference are you expecting? It's going to take a long time to manually breakdown a large request which is why I chose to use the short version to identify differences. If there's a particular thing you're looking for here hints would be good.

If the only difference is string representation I doubt that's something that can be fixed in the driver. ODBC is using a more verbose format and is sending smaller packets so it really wouldn't make sense for it to be quicker than sqlclient.

**Comment by KristianWedberg** [comment]
Is there any way to profile the SQL Server side to see what it's spending all that CPU time on when statements with many parameters come in (as https://github.com/dotnet/SqlClient/issues/974#issuecomment-798778185 shows, that's a very clear difference between SqlClient and Odbc in how the server behaves)?

**Comment by v-chojas** [comment]
It should switch over to a different type because regular nvarchar is limited to 4000 chars. If you post hexdumps of the packets (hex+ASCII) I can tell what's happening pretty easily. What type it's using and how it's sending it (PLP chunks can vary in size, but ODBC will send the parameter declaration list in one chunk) would give a clue about the perf difference.

**Comment by Wraith2** [comment]
[sql traces.zip](https://github.com/dotnet/SqlClient/files/6166098/sql.traces.zip)

**Comment by v-chojas** [comment]
ODBC sends SQL and parameter declaration list as nvarchar(max):

```
0b70   37 00 2c 00 40 00 50 00 32 00 30 00 34 00 38 00   7.,.@.P.2.0.4.8.
0b80   29 00 00 00 00 00 00 00*E7 FF FF 09 04 D0 00 00*   )...............
0b90   fe ff ff ff ff ff ff ff 64 04 00 00 40 00 50 00   ........d...@.P.
0ba0   31 00 20 00 69 00 6e 00 74 00 2c 00 40 00 50 00   1. .i.n.t.,.@.P.
```

 SqlClient sends SQL and parameter declaration list as ntext:

```
1d70   2c 00 40 00 50 00 32 00 30 00 34 00 38 00 29 00   ,.@.P.2.0.4.8.).
1d80   00 00*63 58 A7 00 00 09 04 d0 00 00*58 a7 00 00   ..cX........X...
1d90   40 00 50 00 31 00 20 00 69 00 6e 00 74 00 2c 00   @.P.1. .i.n.t.,.
1da0   40 00 50 00 32 00 20 00 69 00 6e 00 74 00 2c 00   @.P.2. .i.n.t.,.
```

The ntext is an old datatype that is deprecated, there is probably an internal conversion on the server side which causes the performance difference.

**Comment by Wraith2** [comment]
I was trying to work out what 63 was and where it was being written in the driver. I'll have to see if I can unknot the spaghetti code around rpc construction enough to identity and possibly change the data type. If I get it working I'll post another build with some appcontext switches to try out.

**Comment by Wraith2** [comment]
@KristianWedberg ok, give this one a try. All rpc parameters are sent as varchar(length) or varchar(max) depending on the size. netcore only at the moment. There are two appcontext switches you can use the `Switch.Microsoft.Data.SqlClient.UseODBCParameterFormat` one you've already tried and `Switch.Microsoft.Data.SqlClient.OmitNoMetadata` because i'd written it in before we hit this string type info. I don't think the metadata flag is going to do anything but I am interested to see how performance reacts to both settings for parameter names and the nvarchar type.

As usual. clear old versions out of the nuget cache drop in this one and force rebuild.
[Microsoft.Data.SqlClient.3.0.0-dev.zip](https://github.com/dotnet/SqlClient/files/6166795/Microsoft.Data.SqlClient.3.0.0-dev.zip)

**Comment by David-Engel** [comment]
@Wraith2 I got an exception when parameters hit 512:

```
// **************************
// Benchmark: BatchInsert.InsertAsync: ShortRun(IterationCount=3, LaunchCount=1, WarmupCount=3) [providerAsync=MicrosoftSync, rowsPerStatement=1, columnsPerRow=512, numberOfStatements=67, argument=]
// *** Execute ***
// Launch: 1 / 1
// Execute: dotnet "f7f2e8ad-0b34-455d-8860-54c3eea969e0.dll" --benchmarkName "SqlBatchInsertPerformance.BatchInsert.InsertAsync(providerAsync: MicrosoftSync, rowsPerStatement: 1, columnsPerRow: 512, numberOfStatements: 67, argument: )" --job "ShortRun" --benchmarkId 9 in C:\Users\dengel\Downloads\sql-batch-insert-performance-main\bin\Release\net5.0\f7f2e8ad-0b34-455d-8860-54c3eea969e0\bin\Release\netcoreapp5.0
// BeforeAnythingElse

// Benchmark Process Environment Information:
// Runtime=.NET Core 5.0.4 (CoreCLR 5.0.421.11614, CoreFX 5.0.421.11614), X64 RyuJIT
// GC=Concurrent Workstation
// Job: ShortRun(IterationCount=3, LaunchCount=1, WarmupCount=3)

OverheadJitting  1: 1 op, 366900.00 ns, 366.9000 us/op

System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> Microsoft.Data.SqlClient.SqlException (0x80131904): The incoming tabular data stream (TDS) remote procedure call (RPC) protocol stream is incorrect. Parameter 1 (""): Data type 0xE7 has an invalid data length or metadata length.
   at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at Microsoft.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at Microsoft.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString, Boolean isInternal, Boolean forDescribeParameterEncryption, Boolean shouldCacheForAlwaysEncrypted)
   at Microsoft.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean isAsync, Int32 timeout, Task& task, Boolean asyncWrite, Boolean inRetry, SqlDataReader ds, Boolean describeParameterEncryptionRequest)
   at Microsoft.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, TaskCompletionSource`1 completion, Int32 timeout, Task& task, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry, String method)
   at Microsoft.Data.SqlClient.SqlCommand.InternalExecuteNonQuery(TaskCompletionSource`1 completion, Boolean sendToPipe, Int32 timeout, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry, String methodName)
   at Microsoft.Data.SqlClient.SqlCommand.<ExecuteNonQuery>b__143_0()
   at Microsoft.Data.SqlClient.SqlRetryLogicProvider.Execute[TResult](Object sender, Func`1 function)
   at Microsoft.Data.SqlClient.SqlCommand.ExecuteNonQuery()
   at SqlBatchInsertPerformance.BatchInsert.InsertAsync(ProviderAsync providerAsync, Int32 rowsPerStatement, Int32 columnsPerRow, Int32 numberOfStatements, Argument argument) in C:\Users\dengel\Downloads\sql-batch-insert-performance-main\BatchInsert.cs:line 160
   at BenchmarkDotNet.Autogenerated.Runnable_9.WorkloadActionNoUnroll(Int64 invokeCount) in C:\Users\dengel\Downloads\sql-batch-insert-performance-main\bin\Release\net5.0\f7f2e8ad-0b34-455d-8860-54c3eea969e0\f7f2e8ad-0b34-455d-8860-54c3eea969e0.notcs:line 8541
   at BenchmarkDotNet.Engines.Engine.RunIteration(IterationData data)
   at BenchmarkDotNet.Engines.EngineFactory.Jit(Engine engine, Int32 jitIndex, Int32 invokeCount, Int32 unrollFactor)
   at BenchmarkDotNet.Engines.EngineFactory.CreateReadyToRun(EngineParameters engineParameters)
   at BenchmarkDotNet.Autogenerated.Runnable_9.Run(IHost host, String benchmarkName) in C:\Users\dengel\Downloads\sql-batch-insert-performance-main\bin\Release\net5.0\f7f2e8ad-0b34-455d-8860-54c3eea969e0\f7f2e8ad-0b34-455d-8860-54c3eea969e0.notcs:line 7763
ClientConnectionId:d8de5db4-60d1-4b00-b365-92a0041914e8
Error Number:8016,State:30,Class:16
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at BenchmarkDotNet.Autogenerated.UniqueProgramName.AfterAssemblyLoadingAttached(String[] args) in C:\Users\dengel\Downloads\sql-batch-insert-performance-main\bin\Release\net5.0\f7f2e8ad-0b34-455d-8860-54c3eea969e0\f7f2e8ad-0b34-455d-8860-54c3eea969e0.notcs:line 57
// AfterAll
ExitCode != 0
// Benchmark Process 35124 has exited with code -1
```

**Comment by Wraith2** [comment]
Interesting. So how do you specify an unbounded nvarchar?
I thought this this:
```csharp
            sqlParam.SqlDbType = ((paramList.Length << 1) <= TdsEnums.TYPE_SIZE_LIMIT) ? SqlDbType.NVarChar : SqlDbType.NText;
            sqlParam.Value = paramList;
            sqlParam.Size = paramList.Length;
```

would become this:
```csharp
            sqlParam.SqlDbType = SqlDbType.NVarChar;
            sqlParam.Value = paramList;
            sqlParam.Size = ((paramList.Length << 1) <= TdsEnums.TYPE_SIZE_LIMIT) ? paramList.Length : -1;
```

**Comment by David-Engel** [comment]
> perhaps you can only send varchar(max) in plp mode?

@Wraith2 
I think that's correct and it looks like that's what the JDBC driver does:
https://github.com/microsoft/mssql-jdbc/blob/dev/src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java#L4615

**Comment by KristianWedberg** [comment]
Now with the latest MDS 3.0.0-dev, and 256 parameters.

- `OmitNoMetadata` makes no difference, with or without `UseODBCParameterFormat`.
- The latest `varchar` version provides no additional improvement at 256 parameters (379,143 vs. the old 374,091 parameters/s with `UseODBCParameterFormat=true`).

This should obviously be tested with >256 parameters (I got the exception to), which showed a large slowdown with the non-varchar version.

Because of the higher performance, I'll start running with a larger total volume, so these throughput numbers are not directly comparable with previous tests. Again, the interesting bit is if >256 parametes will slow down or not.

| UseODBCParameterFormat | provider | RPS | CPR | Throughput [params/s] |
|--- | ---: | ---: | ---: | ---: |
|  |      Odbc | 16 | 16 | 557,563 |
| true  | Microsoft 3.0.0-dev | 16 | 16 | 625,339 |
| false | Microsoft 3.0.0-dev | 16 | 16 | 218,018 |
|  |    System | 16 | 16 | 215,035 |

**Comment by Wraith2** [comment]
I'm trying it but it isn't going well. The MetaType for varchar isn't PLP so i'm going to end up looking for size==-1 all over the place and even then it's going to touch a whole lot of places i'm really not happy about possibly breaking. I can only do the sync version because the async code in `TDSExecuteRPCAddParameter` means you can't add any data after you've written the parameter so i don't know how to get the PLP chunk terminator written.

**Comment by Wraith2** [comment]
@KristianWedberg  So am I reading the results right that the combination of omitting parameter names and using nvarchar gives better performance than ODBC?

**Comment by KristianWedberg** [comment]
MDS 3.0.0-dev + Omitting parameter names is consistently 5-10% faster than Odbc **_at 256 parameters_**.

`nvarchar` might or might not help a few percent, hard to say since I can't test them both at the same time.

**Comment by Wraith2** [comment]
You can test longer with this. Horrible hackjob, absolutely don't use it for anything other than this testing and only in sync mode or it'll crash like before. 
[Microsoft.Data.SqlClient.3.0.0-dev.zip](https://github.com/dotnet/SqlClient/files/6167388/Microsoft.Data.SqlClient.3.0.0-dev.zip)

**Comment by KristianWedberg** [comment]
Success! :fireworks:

A quick test shows that the latest MDS (and with both switches enabled) scales well all the way to 2048 parameters (at 1,184,713 parameters/s), staying 5-10% faster than Odbc all the way! Full results will take longer.

**Note**: I did realize I had one issue in my testing setup until now: I was still giving MDS less work (since it used to be up to 20x slower). It might not have been the latest change that did the trick, it could e.g. be just the `UseODBCParameterFormat` switch, I'll go back and check this now.

**Comment by Wraith2** [comment]
OK. Well consider that a proof of concept and now we know it's possible to make it work it'll need to be worked on to make it production capable. I need to learn how the driver is currently handling varchar(max) because it's clearly using some trickery if I've had to add workarounds to make it send a long string correctly. This could take some time.

**Comment by KristianWedberg** [comment]
To sum up the performance measurements:

- There is no performance difference between the latest MDS `nvarchar` hackjob and the earlier MDS `UseODBCParameterFormat` without `nvarchar` (now that I'm running with the same data volume), i.e. the `nvarchar` hackjob changes are _not_ needed to get great performance.
- With the latest MDS `nvarchar` hackjob, enabling/disabling `OmitNoMetadata` makes no difference, from 1 to 2048 parameters.
- Disabling `UseODBCParameterFormat` makes MDS slow down at >256 parameters.

Here are repeatable results using the earlier `UseODBCParameterFormat` MDS (i.e. _without_ `nvarchar` changes):

| UseODBCParameterFormat | provider | RPS | CPR | Parameters | Throughput [params/s] |
|--- | ---: | ---: | ---: | ---: | ---: |
|       |    System | 32 | 8 | 256 | 222,754 |
|       |    System | 32 | 16 | 512 | 157,327 |
|       |    System | 32 | 32 | 1024 | 90,484 |
|       |    System | 32 | 64 | 2048 | 44,815 |
| false | Microsoft | 32 | 8 | 256 | 230,582 |
| false | Microsoft | 32 | 16 | 512 | 158,538 |
| false | Microsoft | 32 | 32 | 1024 | 92,574 |
| false | Microsoft | 32 | 64 | 2048 | 49,096 |
| true | Microsoft | 32 | 8 | 256 | 600,829 |
| true | Microsoft | 32 | 16 | 512 | 801,047 |
| true | Microsoft | 32 | 32 | 1024 | 1,080,539 |
| true | Microsoft | 32 | 64 | 2048 | 1,239,986 |
|       |    Odbc | 32 | 8 | 256 | 549,672 |
|       |    Odbc | 32 | 16 | 512 | 755,040 |
|       |    Odbc | 32 | 32 | 1024 | 984,802 |
|       |    Odbc | 32 | 64 | 2048 | 1,120,003 |

So performance looks great. 

Functionality wise, I'll highlight that my benchmark only uses each parameter once, in the same order they're written in the query. I'm guessing reusing a named parameter multiple times in the query requires additional testing.

**Comment by Wraith2** [comment]
> •	There is no performance difference between the latest MDS nvarchar hackjob and the earlier MDS UseODBCParameterFormat without nvarchar (now that I'm running with the same data volume), i.e. the nvarchar hackjob changes are not needed to get great performance.

Oh, that's surprising but welcome. 
The parameter format is pretty easy it just needs some agreement on how to implement it so you can use it on a per-command basis. I listed 3 strategies earlier in the thread. That'll be down to @David-Engel and @cheenamalhotra 

> I'm guessing reusing a named parameter multiple times in the query requires additional testing. 

yes, definitely. I expect it not to work.

> Disabling UseODBCParameterFormat makes MDS slow down at >256 parameters.

That's got to be down to how the name mapping is implemented on the server. I doubt anyone expected it to reach the limit of parameters, that might be something @v-chojas could explore. As I think I said earlier the falloff curve looks list iteration instead of map lookup which makes sense because list tends to be faster than map when items<~100.

**Comment by David-Engel** [comment]
I'm not able to replicate the improved test results with the latest MDS package (hackjob) from @Wraith2 or the previous package. I've triple checked my MDS DLL and I'm pretty sure I'm using the correct one (modified dates line up to the correct times in my build output) and I've tried with and without the UseODBCParameterFormat switch. Performance still falls off around 256 params on all runs I try. 🤔 

@KristianWedberg Can you push any changes you made to the test to your [repo](https://github.com/KristianWedberg/sql-batch-insert-performance)?

**Comment by KristianWedberg** [comment]
@David-Engel Yes, that will be due to running with too small a data volume for the fast MDS. I've refreshed the [benchmark code](https://github.com/KristianWedberg/sql-batch-insert-performance) to address that (for me each test case now takes about 1 second). You'll have to upgrade from MDS 2.1.2.
